File: packages/scoring/definitions/pss.ts
/**
 * Perceived Stress Scale — 10-item version (PSS-10) — SurveyJS Schema + Scoring Rules
 * ======================================================================================
 * Cohen, Kamarck & Mermelstein (1983). Journal of Health and Social Behavior, 24, 385–396.
 *
 * 10 items. Items 4, 5, 7, 8 are positively worded and must be reverse-scored.
 * Scale: 0 (Never) → 4 (Very Often). Total range: 0–40.
 */

const PSS_CHOICES = [
  { value: 0, text: "Never" },
  { value: 1, text: "Almost Never" },
  { value: 2, text: "Sometimes" },
  { value: 3, text: "Fairly Often" },
  { value: 4, text: "Very Often" },
];

// Positive items displayed with the same scale but reverse-scored during computation
const PSS_CHOICES_POSITIVE = PSS_CHOICES; // displayed identically; reversal done in scoring

export const PSS_SURVEY_SCHEMA = {
  title: "Perceived Stress Scale (PSS-10)",
  description:
    "The questions in this scale ask you about your feelings and thoughts during the last month. In each case, please indicate how often you felt or thought a certain way.",
  logoPosition: "right",
  showProgressBar: "top",
  showQuestionNumbers: "on",
  questionsOnPageMode: "singlePage",
  pages: [
    {
      name: "pss",
      title: "In the last month, how often have you…",
      elements: [
        {
          type: "radiogroup",
          name: "pss_1",
          title: "…been upset because of something that happened unexpectedly?",
          choices: PSS_CHOICES,
          isRequired: true,
        },
        {
          type: "radiogroup",
          name: "pss_2",
          title: "…felt that you were unable to control the important things in your life?",
          choices: PSS_CHOICES,
          isRequired: true,
        },
        {
          type: "radiogroup",
          name: "pss_3",
          title: "…felt nervous and stressed?",
          choices: PSS_CHOICES,
          isRequired: true,
        },
        {
          type: "radiogroup",
          name: "pss_4",
          title: "…felt confident about your ability to handle your personal problems?",
          description: "This item is reverse-scored.",
          choices: PSS_CHOICES_POSITIVE,
          isRequired: true,
        },
        {
          type: "radiogroup",
          name: "pss_5",
          title: "…felt that things were going your way?",
          description: "This item is reverse-scored.",
          choices: PSS_CHOICES_POSITIVE,
          isRequired: true,
        },
        {
          type: "radiogroup",
          name: "pss_6",
          title: "…found that you could not cope with all the things that you had to do?",
          choices: PSS_CHOICES,
          isRequired: true,
        },
        {
          type: "radiogroup",
          name: "pss_7",
          title: "…been able to control irritations in your life?",
          description: "This item is reverse-scored.",
          choices: PSS_CHOICES_POSITIVE,
          isRequired: true,
        },
        {
          type: "radiogroup",
          name: "pss_8",
          title: "…felt that you were on top of things?",
          description: "This item is reverse-scored.",
          choices: PSS_CHOICES_POSITIVE,
          isRequired: true,
        },
        {
          type: "radiogroup",
          name: "pss_9",
          title: "…been angered because of things that were outside of your control?",
          choices: PSS_CHOICES,
          isRequired: true,
        },
        {
          type: "radiogroup",
          name: "pss_10",
          title:
            "…felt difficulties were piling up so high that you could not overcome them?",
          choices: PSS_CHOICES,
          isRequired: true,
        },
      ],
    },
  ],
};

export const PSS_SCORING_RULES = {
  type: "PSS",
  version: "1.0",
  items: {
    negative: ["pss_1","pss_2","pss_3","pss_6","pss_9","pss_10"],
    positive: ["pss_4","pss_5","pss_7","pss_8"],  // reverse-scored
  },
  total: {
    method: "sum",
    range: { min: 0, max: 40 },
  },
  bands: [
    { min: 0,  max: 13, label: "Low",      description: "Low perceived stress." },
    { min: 14, max: 26, label: "Moderate", description: "Moderate perceived stress. Monitor stress levels and use coping strategies." },
    { min: 27, max: 40, label: "High",     description: "High perceived stress. Consider speaking with a healthcare professional." },
  ],
};

File: packages/scoring/pss.py
"""
Perceived Stress Scale — 10-item version (PSS-10) — Scoring Module
===================================================================
Reference: Cohen, S., Kamarck, T., & Mermelstein, R. (1983). A global measure
of perceived stress. Journal of Health and Social Behavior, 24, 385-396.

Items:
  Negative (scored as-is): 1, 2, 3, 6, 9, 10
  Positive (reverse scored): 4, 5, 7, 8

Response scale (caller passes raw 0–4 integer):
  0 = Never
  1 = Almost Never
  2 = Sometimes
  3 = Fairly Often
  4 = Very Often

Total score: sum of all 10 items (after reversing positives), range 0–40.

Interpretation bands:
  0–13   → Low stress
  14–26  → Moderate stress
  27–40  → High stress
"""

from __future__ import annotations
from dataclasses import dataclass, field
from typing import Dict, List

NEGATIVE_ITEMS = {f"pss_{i}" for i in [1, 2, 3, 6, 9, 10]}
POSITIVE_ITEMS = {f"pss_{i}" for i in [4, 5, 7, 8]}   # reverse-scored
ALL_ITEMS      = [f"pss_{i}" for i in range(1, 11)]

VALID_VALUES = {0, 1, 2, 3, 4}

BANDS = [
    (0,  13, "Low"),
    (14, 26, "Moderate"),
    (27, 40, "High"),
]


@dataclass
class PSSResult:
    total_score: int
    band: str

    def to_dict(self) -> dict:
        return {
            "total": {
                "score": self.total_score,
                "band": self.band,
            }
        }


def _get_band(score: int) -> str:
    for low, high, label in BANDS:
        if low <= score <= high:
            return label
    return "High"


def _reverse(value: int) -> int:
    """Reverse 0–4 Likert value: 0↔4, 1↔3, 2↔2."""
    return 4 - value


def validate_responses(responses: Dict[str, int]) -> List[str]:
    errors: List[str] = []
    missing = [k for k in ALL_ITEMS if k not in responses]
    if missing:
        errors.append(f"Missing items: {missing}")
    for key, val in responses.items():
        if key in NEGATIVE_ITEMS | POSITIVE_ITEMS:
            if val not in VALID_VALUES:
                errors.append(f"Item {key} has invalid value {val}. Must be 0–4.")
    return errors


def score(responses: Dict[str, int]) -> PSSResult:
    """
    Compute PSS-10 total score.

    Parameters
    ----------
    responses : dict
        Mapping of item key → integer value (0–4).
        e.g. {"pss_1": 2, "pss_2": 3, "pss_4": 1, …}

    Returns
    -------
    PSSResult with total score and interpretation band.
    """
    errors = validate_responses(responses)
    if errors:
        raise ValueError(f"PSS validation failed: {errors}")

    total = 0
    for key in ALL_ITEMS:
        val = responses[key]
        total += _reverse(val) if key in POSITIVE_ITEMS else val

    return PSSResult(total_score=total, band=_get_band(total))
